### PR TITLE
bringing karma for browser testing

### PR DESCRIPTION
**Pull request by Jabher**
So, let's talk one more time about testing.

This PR changes the following:
- adds karma test runner (over existing mocha suite and dist version of synaptic)
- adds multiple npm scripts to run mocha against src and dist versions
- defining separate travis test runner which is running mocha against both src and dist versions
- raises timeout for networks comparison tests to 30 seconds (as they take a significant amount of time on some not very powerful machines)
- raises threshold level for "Optimized and Unoptimized Networks Equivalency" test to 1e-9 as I'm actually tired of getting "got 1.07e-10, however, expected 1e-10" error message.

Node.js testing is great, but we possibly need to assure that everything is working as expected for built `.js` file too and possibly run the tests against browsers (to make sure that everything is working not only in node.js).

I cannot integrate browser testing (there's one free-for-opensource SaaS) with travis CI, but at least local karma testing is possible. 
Actually it is possible to have some kind of badge like the one in [this repo](https://github.com/vuejs/vue) in future, but I did never work with sauce labs.

### CODE REVIEW

**Comment by cazala** [comment]
Sweet! We could use Karma + Chrome in Travis, would you like to add it to the PR? Just add this to `.travis.yml`

```
before_install:
- export CHROME_BIN=chromium-browser
- export DISPLAY=:99.0
- sh -e /etc/init.d/xvfb start
- sleep 3
```

And then this to `karma.config.js`:

```javascript
var configuration = { ...karma config... }
if (process.env.TRAVIS) {
  configuration.browsers = ['Chrome_travis_ci']
  configuration.customLaunchers = {
    Chrome_travis_ci: {
      base: 'Chrome',
      flags: ['--no-sandbox']
    }
  }
}
module.exports = configuration
```

And then just call `karma` without passing `--browsers` from the `test:travis` script

**Comment by menduz** [comment]
@Jabher 👆

**Comment by Jabher** [comment]
Emm, seems like it's not working. I have a feeling that since Travis integrated with sauce labs they dropped browsers support

File: karma.conf.js [hunk 1]
```diff
@@ -21,5 +21,17 @@ module.exports = function(config) {
     singleRun: false,
     concurrency: Infinity,
     browserNoActivityTimeout: 60000,
-  })
-}
+  });
+
+  if (process.env.TRAVIS) {
+    config.set({
+      browsers: ['Chrome_travis_ci', 'Firefox'],
```
Comment (cazala) [inline karma.conf.js]: I think travis doesn't support Firefox

File: package.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
     "test:mocha:dist": "mocha test --require dist/synaptic.js ./test",
     "test:karma:browsers": "karma start --single-run --browsers Chrome,Firefox,SafariPrivate",
     "test:karma:phantomjs": "karma start --single-run --browsers PhantomJS",
-    "test:travis": "npm run test:mocha:src && npm run build && npm run test:mocha:dist",
+    "test:karma:travis": "TRAVIS=true karma start --single-run",
```
Comment (cazala) [inline package.json]: Travis sets the TRAVIS env var by itself

**Comment by cazala** [comment]
Okay, feel free to rollback the changes. I'll see if it is possible to run the tests with Travis + Chrome, I'm pretty sure we do that here at work so there must be a way, but I think it is okay to merge what you did already (:

**Comment by Jabher** [comment]
Well, speaking about FF - travis supported it, at least somewhen in the past: https://docs.travis-ci.com/user/firefox/

I definitely feel that there's no browsers in travis right now.

**Comment by cazala** [comment]
I got something working with Karma + Chrome + Travis. I ll merge this and see if I can figure out a way to make it work. I'm using travis-ci**.com** tho (the paid one, not **.org**) maybe that makes a difference..

dummy test

**Comment by Jabher** [comment]
Cool! i've encountered some mentions of the fact that .com and .org are a bit different, but did not expect something of that kind.

**Pull request merged**